src/cudamatrix/cu-kernels.cu

```diff
@@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
   }
 }
 
+template<typename Real>
+__global__
+static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
+                                  Real* obfj_terms, MatrixDim objf_dim, 
+                                  Real* obfj_derivs, MatrixDim derivs_dim) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
+  int32_cuda scores_index = i + j * scores_dim.stride;
+  Real K = 1.0 / (scores_dim.rows - 2.0);
+  Real L = scores[scores_index];
+  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
```

Comment:
OK, let me work this out...
The forward computation is something like:

A = X X'
cvec = diag(X S X')
u = vector of ones
S = A - cvec u' - u cvec + b
... compute the objf and get S_deriv which is d(objf)/dS
A_deriv = S_deriv
X_deriv += 2 A_deriv X   (or something like that)
cvec_deriv = - sum-of-Sderiv-cols - sum-of-Sderiv-rows
when computing the deriv w.r.t. S I am thinking about the expression
cvec_deriv . cvec,
which equals trace(diag(cvec_deriv) X S X'), where diag(cvec_deriv) is a
matrix whose diagonal is cvec_deriv, which we can rearrange to trace(S (X'
diag(cvec_deriv) X)).
We get from this (through a mysterious process, I do it intuitively),
S_deriv = X' diag(cvec_deriv) X'
which is pretty easy to compute.







On Sat, Feb 13, 2016 at 5:09 PM, david-ryan-snyder <notifications@github.com
> wrote:

> In src/cudamatrix/cu-kernels.cu
> <https://github.com/danpovey/kaldi/pull/5#discussion_r52834379>:
>
> > @@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
> >    }
> >  }
> >
> > +template<typename Real>
> > +__global__
> > +static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
> > +                                  Real* obfj_terms, MatrixDim objf_dim,
> > +                                  Real* obfj_derivs, MatrixDim derivs_dim) {
> > +  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
> > +  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
> > +  int32_cuda scores_index = i + j * scores_dim.stride;
> > +  Real K = 1.0 / (scores_dim.rows - 2.0);
> > +  Real L = scores[scores_index];
> > +  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
>
> After that you just have to do the reverse of
> the forward operations to get the derivatives w.r.t. S and the matrix of
> xvectors.
>
> Right, that's what I'm referring to. Once you have the derivs of the objf
> wrt to the scores, you still need to compute the derivative of the scores
> wrt to S. However, as far as I can tell, unless you try to do that in a
> kernel, you'll end up with an algorithm with two loops over the xvectors. I
> proposed the kernel above to parallelize that computation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/5/files#r52834379>.
>